initfiles/bin/init_thor

```diff
@@ -15,26 +15,195 @@
 
-export deploydir=$(dirname $(type -path $0)) 
-PATH_PRE=`type -path hpcc_setenv` 
+deploydir=$(dirname $(type -path $0))
+PATH_PRE=$(type -path hpcc_setenv)
 source ${PATH_PRE} 
-PID_NAME="$PID/`basename $PWD`.pid"
 
-INIT_PID_NAME="$PID/init_`basename $PWD`.pid"
+INSTALL_DIR=$(dirname ${PATH_PRE})/..
+source  ${INSTALL_DIR}/etc/init.d/hpcc_common
+
+PID_NAME="$PID/$(basename $PWD).pid"
+INIT_PID_NAME="$PID/init_$(basename $PWD).pid"
 echo $$ > $INIT_PID_NAME
 
 export SENTINEL="thor.sentinel"
 rm -f ${SENTINEL}
 
-killed() {
-        echo "Stopping"
-        kill_process ${SENTINEL} ${PID_NAME} 3
-        exit 255
+if [[ -z "$deploydir" ]]; then
+    deploydir=$(pwd -P)
+fi
+
+compname=$(basename $PWD)
+instancedir=$(pwd -P)
+source $instancedir/setvars
+
+if [[ ! -z ${THORPRIMARY} ]]; then
+    groupName=${THORPRIMARY}
+else
+    groupName=${THORNAME}
+fi
+ln -s -f $deploydir/thormaster${LCR} thormaster_$THORNAME
+
+ENV_DIR=$(cat ${HPCC_CONFIG} | sed -n "/\[DEFAULT\]/,/\[/p" | grep "^configs=" | sed -e 's/^configs=//')
+logdir=$(updtdalienv $ENV_DIR/environment.xml -d log thor $THORNAME)
+logfile=$logdir/${THORNAME}.log
+
+contains()
+{
+    local n=$#
+    local value=${!n}
+    local i=0
+    for ((i=1;i < ${n};i++)); do
+        if [[ "${!i}" == "${value}" ]]; then
+            echo "${i}"
+            return 0
+        fi
+    done
+    echo "0"
+    return 1
+}
+
+makethorgroup()
+{
+    if [[ -z "$localthorportinc" ]]; then
+        localthorportinc=200
+    fi
+    rm -f thorgroup
+
+    declare -a ports_used
+    declare -a hosts
+    for slave in $(cat slaves); do
+        p=$(contains "${hosts[@]}" "${slave}")
+        if [[ 0 == ${p} ]]; then
+            echo "${slave}:${THORSLAVEPORT}" >> thorgroup
+            p=$((${#hosts[@]}+1))
+            ports[${p}]=${THORSLAVEPORT}
+            hosts=(${hosts[@]} $slave)
+        else
+            newport=$((${ports[${p}]}+${localthorportinc}))
+            echo "${slave}:${newport}" >> thorgroup
+            ports[${p}]=${newport}
+        fi
+    done
+}
+
+kill_slaves()
+{
+    if [[ "$localthor" = "true" ]]; then
+        killall -9 thorslave_${compname} 2> /dev/null
+    else
+        # we want to kill only slaves that have already been started in run_thor
+        if [[ -r $instancedir/uslaves.start ]]; then
+            nslaves=$(cat $instancedir/uslaves.start 2> /dev/null | wc -l)
+            $deploydir/frunssh $instancedir/uslaves.start "/bin/sh -c '$deploydir/init_thorslave %a $THORMASTER $THORMASTERPORT $logdir $instancedir $deploydir $THORNAME $PATH_PRE $logredirect stop'" -i:$SSHidentityfile -u:$SSHusername -pe:$SSHpassword -t:$SSHtimeout -a:$SSHretries -n:$nslaves 2>&1 | egrep -v "no process killed"
+            echo slaves stopped
+        fi
+    fi
+
+    rm -f $instancedir/uslaves.start > /dev/null 2>&1
+}
+
+killed()
+{
+    echo "Stopping"
+    kill_process ${PID_NAME} thormaster_${compname} 8
+    kill_slaves
```

Comment:
So this line is a modification that I recently did.  The thinking behind it is that if at all possible, we want kill_process to send a sigterm to thormaster and have it spin down it's slaves, without us having to take a hammer (kill_slaves) to it first.  Before we were just doing a killall -9 on the slaves, which would cause thormaster to attempt to respawn the slaves.  If done after the kill_process call, we guarantee that we clean up any remnants of the thor cluster instead of just forcing the master to do extra work before we shut it down.